apps/prairielearn/src/pages/instructorAssessmentStatistics/instructorAssessmentStatistics.html.ts

```diff
@@ -0,0 +1,333 @@
+  threshold_seconds: AssessmentSchema.shape.duration_stat_threshold_seconds,
+  threshold_labels: AssessmentSchema.shape.duration_stat_threshold_labels,
+  hist: AssessmentSchema.shape.duration_stat_hist,
+});
+export type DurationStat = z.infer<typeof DurationStatSchema>;
+
+export const AssessmentScoreHistogramByDateSchema = z.object({
+  date: z.date(),
+  date_formatted: z.string(),
+  number: z.string(),
+  mean_score_perc: z.number(),
+  histogram: z.array(z.number()),
+});
+type AssessmentScoreHistogramByDate = z.infer<typeof AssessmentScoreHistogramByDateSchema>;
+
+export const UserScoreSchema = z.object({
+  duration_secs: z.number(),
+  score_perc: AssessmentInstanceSchema.shape.score_perc,
+});
+type UserScore = z.infer<typeof UserScoreSchema>;
+
+export function InstructorAssessmentStatistics({
+  resLocals,
+  durationStat,
+  assessmentScoreHistogramByDate,
+  userScores,
+}: {
+  resLocals: Record<string, any>;
+  durationStat: DurationStat;
+  assessmentScoreHistogramByDate: AssessmentScoreHistogramByDate[];
+  userScores: UserScore[];
+}) {
+  return html`
+    <!doctype html>
+    <html lang="en">
+      <head>
+        ${renderEjs(__filename, "<%- include('../partials/head'); %>", {
+          ...resLocals,
+          pageTitle: 'Statistics',
+        })}
+        <script src="${nodeModulesAssetPath('lodash/lodash.min.js')}"></script>
+        <script src="${nodeModulesAssetPath('d3/dist/d3.min.js')}"></script>
+        <script src="${assetPath('localscripts/histogram.js')}"></script>
+        <script src="${assetPath('localscripts/scatter.js')}"></script>
+        <script src="${assetPath('localscripts/parallel_histograms.js')}"></script>
+      </head>
+      <body>
+        ${renderEjs(__filename, "<%- include('../partials/navbar'); %>", resLocals)}
+        <main id="content" class="container-fluid">
+          ${renderEjs(
+            __filename,
+            "<%- include('../partials/assessmentSyncErrorsAndWarnings'); %>",
+            resLocals,
+          )}
+
+          <div class="card mb-4">
+            <div class="card-header bg-primary text-white">
+              ${resLocals.assessment_set.name} ${resLocals.assessment.number}: Score statistics
+            </div>
+            ${resLocals.assessment.score_stat_number > 0
+              ? html`
+                  <div class="card-body">
+                    <div id="scoreHist" class="histogram"></div>
+                    <script>
+                      $(function () {
+                        var data = [${escapeHtml(resLocals.assessment.score_stat_hist)}];
+                        var xgrid = _.range(0, 110, 10);
+                        var options = {
+                          ymin: 0,
+                          xlabel: 'score / %',
+                          ylabel: 'number of students',
+                        };
+                        histogram('#scoreHist', data, xgrid, options);
+                      });
+                    </script>
```

Comment:
We have been moving away from dynamic JS code generation. Here's a suggested approach:
```suggestion
                    <div id="scoreHist" class="histogram" data-histogram="${JSON.stringify(resLocals.assessment.score_stat_hist)}"></div>
                    <script>
                      $(function () {
                        const data = JSON.parse(document.getElementById("scoreHist").dataset.histogram);
                        const xgrid = _.range(0, 110, 10);
                        const options = {
                          ymin: 0,
                          xlabel: 'score / %',
                          ylabel: 'number of students',
                        };
                        histogram('#scoreHist', data, xgrid, options);
                      });
                    </script>
```

Same for all references to dynamic data in scripts. Even better if this can be moved to an asset, but I won't block on that.

Reply:
In [fe614f8](https://github.com/PrairieLearn/PrairieLearn/pull/9809/commits/fe614f896046789604dac2796569dc2b7421cebc), added `instructorAssessmentStatisticsClient` where we are handling this and picking data attributes off of the elements.